File: solution/148SortList.cpp
//
// Created by Alaric on 2019-10-05.
//
//top-down
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* sortList(ListNode* head) {
        if (!head || !head->next) return head;
        ListNode* slow = head;
        ListNode* fast = head->next;
        while(fast && fast->next){
            slow = slow->next;
            fast = fast->next->next;
        }
        ListNode* mid = slow->next;
        slow->next = nullptr;
        return merge(sortList(head) , sortList(mid));
    }
private:
    ListNode* merge(ListNode* l1 , ListNode* l2) {
        ListNode dummy(0);
        ListNode* tail = &dummy;
        while (l1 && l2){
            if (l1->val > l2->val) swap(l1,l2);
            tail->next = l1;
            l1 = l1->next;
            tail = tail->next;
        }
        if (l1) tail->next = l1;
        if (l2) tail->next = l2;
        return dummy.next;
    }
};

File: solution/148SortList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
# 会超时
class Solution:
    def sortList(self, head: ListNode) -> ListNode:
        def quicksort(head,end):
            if head == end or head.next == end:
                return
            p = head.next
            pivot = head.val
            mid = head
            while p != end:
                if p.val <  pivot:
                    mid = mid.next
                    mid.val , p.val = p.val,mid.val
                p = p.next
            head.val = mid.val
            mid.val = pivot
            quicksort(head,mid)
            quicksort(mid.next,end)
            
        quicksort(head,None)
        return head